src/Features/Core/Portable/Completion/Providers/AbstractPartialTypeCompletionProvider.cs

```diff
@@ -0,0 +1,120 @@
+{
+    internal abstract partial class AbstractPartialTypeCompletionProvider : CommonCompletionProvider
+    {
+        protected AbstractPartialTypeCompletionProvider()
+        {
+        }
+
+        public async sealed override Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+            var document = completionContext.Document;
+            var position = completionContext.Position;
+            var cancellationToken = completionContext.CancellationToken;
+
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var node = GetPartialTypeSyntaxNode(tree, position, cancellationToken);
+
+            if (node != null)
+            {
+                var semanticModel = await document.GetSemanticModelForNodeAsync(node, cancellationToken).ConfigureAwait(false);
+                var syntaxContext = await CreateSyntaxContextAsync(document, semanticModel, position, cancellationToken).ConfigureAwait(false);
+
+                var declaredSymbol = semanticModel.GetDeclaredSymbol(node, cancellationToken) as INamedTypeSymbol;
+
+                if (declaredSymbol != null)
+                {
+                    var symbols = LookupCandidateSymbols(syntaxContext, declaredSymbol, cancellationToken);
+                    var items = symbols?.Select(symbol => CreateCompletionItem(symbol, completionContext.DefaultItemSpan, position, syntaxContext));
+
+                    if (items != null)
+                    {
+                        completionContext.AddItems(items);
+                    }
+                }
+            }
+        }
+
+        private CompletionItem CreateCompletionItem(INamedTypeSymbol symbol, TextSpan textSpan, int position, AbstractSyntaxContext context)
+        {
+            var displayAndInsertionText = GetDisplayAndInsertionText(symbol, context);
+
+            return SymbolCompletionItem.Create(
+                displayText: displayAndInsertionText.Item1,
+                insertionText: displayAndInsertionText.Item2,
+                span: textSpan,
+                symbol: symbol,
+                contextPosition: context.Position,
+                descriptionPosition: position,
+                rules: CompletionItemRules.Default);
+        }
+
+        protected abstract Task<AbstractSyntaxContext> CreateSyntaxContextAsync(
+            Document document,
+            SemanticModel semanticModel,
+            int position,
+            CancellationToken cancellationToken);
+
+        protected abstract SyntaxNode GetPartialTypeSyntaxNode(SyntaxTree tree, int position, CancellationToken cancellationToken);
+
+        protected abstract ValueTuple<string, string> GetDisplayAndInsertionText(INamedTypeSymbol symbol, AbstractSyntaxContext context);
+
+        protected virtual IEnumerable<INamedTypeSymbol> LookupCandidateSymbols(AbstractSyntaxContext context, INamedTypeSymbol declaredSymbol, CancellationToken cancellationToken)
```

Comment:
Yes, the C# implementation needs to refine the results to only those with a `partial` modifier.